### ISSUE TITLE
Typescript declarations: children gets type-checked as an attribute?

### ISSUE THREAD

**Issue by mindplay-dk** [developit/preact#1008]
I've run into a peculiar type-checking problem with functional components.

```
const Example = ({ title, children }: { title?: string, children: VNode[] }) => (
    <div class="example">
        <div class="example__source">
            <pre><code class="language-html">{renderHTML(children)}</code></pre>
        </div>
        <div class="example__content">{children}</div>
    </div>
);

const doc = (
    <div>
        <h1>Typography</h1>
        <Example>
            <h1>Heading 1</h1>
            <h2>Heading 2</h2>
            <h3>Heading 3</h3>
        </Example>
    </div>
);
```

Every call to `<Example>` complains that `children` is missing, which obviously isn't true.

I've always thought it was odd that `children`, being completely unrelated to `props`, is mixed into the `props` object, rather than being passed as a second argument - presumably that's what causes this problem? But I guess we inherited this weirdness from React 🙄 

I could of course type-hint as `children?` any time a functional component accepts `children`, making it optional - but in this example (and others) it *isn't* optional, and the the problem just moves to `renderHTML(children)` where I then have to write `renderHTML(children!)` which also isn't strictly correct, or `renderHTML(children || [])` which *would* be correct if `children` was really optional in the first place, so this is even more confusing.

Is there any way to improve the type-hinting of `children` somehow?

I suspect the only way this could ever improve is if the maintainers of React could be convinced to untangle `children` from `props` - which is hardly worth even propositioning to them, as that would break literally every component in every project.

I thought I'd ask anyway 😃

**Comment by developit** [comment]
That's likely due to the definitions being based on React's. We pull `children` out of `props`, but that's actually the original location of it, and might be changing in a future version.

Preact's definition shouldn't have children listed as optional - it's always present and always an Array.

**Comment by mindplay-dk** [comment]
> We pull `children` out of `props`

You do? I've been getting `children` via `props.children`, and that works. Removing it from `props` would be breaking vs React, right? (Which I suppose might be okay in a major release, assuming the React compatibility-package can restore it?)

> Preact's definition shouldn't have children listed as optional - it's always present and always an Array

Good to know.

On that note, I've had `attributes` come up as `null` - I've had to do e.g. `attributes || {}` in several places. It would be nice if this consistently came through as an Object as well, but I suppose that's a (however small) breaking change, too?

**Comment by developit** [comment]
I believe that's how it works now yes - we pull it out, but `preact-compat` adds it back in as an alias within props.

**Comment by mindplay-dk** [comment]
> we pull it out, but `preact-compat` adds it back in as an alias within props.

I don't use `preact-compat`, and I'm getting `children` via `props` - I don't believe the type-hinting includes as separate argument for `children` either?

If the type-hints were corrected in `preact`, likely `preact-compat` would need additional type-hinting, e.g. using declaration merging to extend `preact`'s type declarations.

**Comment by developit** [comment]
Ah - we're talking about 2 different things:

```js
let props = h('a', {}, 'foo').attributes
props.children  // undefined

const Foo = props => {
  props.children  // ["foo"]
}
render(<Foo children={["foo"]}>)
```

**Comment by mindplay-dk** [comment]
Yeah, I'm sorry, I guess I kind of conflated two issues into one here 🙄 

To be clear, the primary issue here is the type-hinting.

I kind of got side-tracked by the fact that P/React passes children via the `props` argument. It was an odd decision on React's part, IMO - children obviously are not properties of elements, and it would conflict with an actual attribute/property named `children`. I strongly prefer what e.g. [this library](https://github.com/jorgebucaran/patchdom/blob/master/patchdom.d.ts#L10) does. It's kind of separate issue though, and I don't know that there's anything we can do about it, since the [render signature](https://github.com/developit/preact/blob/master/src/preact.d.ts#L73) of P/React already uses the second and third arguments for other things.

So ignore that 😃

**Comment by marvinhagemeister** [comment]
@mindplay-dk I'll close this one as the question seems to have been addressed. Happy to reopen if you feel like this is still an issue

**Comment by mindplay-dk** [comment]
@marvinhagemeister well, @developit confirmed the original issue:

> Preact's definition shouldn't have children listed as optional - it's always present and always an Array.

was this addressed?

**Comment by marvinhagemeister** [comment]
This is fixed by #1116 . For preact `children` prop is always an array, but TypeScript treats the `children` prop in its own way. It treats `children` only as an array, when it has more than one child. If it only has one child, it infers the type as being non-array. See the linked PR for more discussion on this topic.
